DO optimization

chenhu-wang opened the pull request:
### Details:
 - *detectionOutput optimization*

### Tickets:
 - *51530*

chenhu-wang commented:
@dmitry-gorokhov ,could you please take a look?

dmitry-gorokhov commented:
@a-sidorova Please review the PR

a-sidorova reviewed (commented):
@chenhu-wang major comment: method `execute` is very large (~450 code lines) and therefore difficult to understand. I suggest to create some methods by meaning and call it from `execute`

chenhu-wang commented:
> @chenhu-wang major comment: method `execute` is very large (~450 code lines) and therefore difficult to understand. I suggest to create some methods by meaning and call it from `execute`

@a-sidorova I have extracted some functions with self-explanation name to be called. Some part is not extracted from perf consideration, and I added comments to help understand. Thank you!

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-                    ymax = (std::max)(0.0f, (std::min)(1.0f, ymax));
+void MKLDNNDetectionOutputNode::confReorderDense(const float *confData, const float *ARMConfData, float *reorderedConfData) {
+    if (withAddBoxPred) {
+        for (int n = 0; n < imgNum; ++n) {
```
Could we use parallelism?

chenhu-wang replied:
yes, done, thank you.

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+    int countIn = detections[0];
+    detections[0] = 0;
+
+    for (int i = 0; i < countIn; ++i) {
```
Could we use parallelism?

chenhu-wang replied:
NMS routine can not paralleled because the next iteration depends on previous iteration which may change the gradually generated output. The next iteration need all current outputs to suppress. Thank you.

a-sidorova replied:
Yes, you're right. Thank you for the explanation!

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-            for (int c = 0; c < _num_classes; ++c) {
-                int detections = detections_data[n*_num_classes + c];
-                int *pindices = indices_data + n*_num_classes*_num_priors + c*_num_priors;
+            for (int c = 0; c < classesNum; ++c) {
```
Could we use parallelism?

chenhu-wang replied:
yes, done, thank you.

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+
+    // set final detection result to output blob
+    int count = 0;
+    for (int n = 0; n < imgNum; ++n) {
+        const float *pconf   = reorderedConfData + n * confInfoLen * classesNum;
+        const float *pboxes  = decodedBboxesData + n * priorsNum * 4 * locNumForClasses;
+        const int *pindices  = indicesData + n * classesNum * priorsNum;
+
+        for (int c = 0; c < classesNum; ++c) {
```
Could we use parallel_2d?

chenhu-wang replied:
The output order matters. We can not paralleled directly. We can parallel by computing the output position offset of each batch and class with traverse all detection outputs, but this add additional overhead, together with this part consuming very small proportion times compare overall time of DO layer, i.e. not heavy workload. so does not give obvious perf benefit in most case. I think we can keep it to make it easy to understand. what do you think. Thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+    if (!isPriorsPerImg && imgNum > 1) {
+        for (int n = 1; n < imgNum; ++n) {
+            numPriorsActualdata[n] = numPriorsActualdata[0];
         }
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.h:
```diff
@@ -15,72 +16,93 @@ class MKLDNNDetectionOutputNode : public MKLDNNNode {
-    std::vector<float> _reordered_conf;
-    std::vector<float> _bbox_sizes;
-    std::vector<int> _num_priors_actual;
+    void confFilterCF(float* reorderedConfData, int* indicesData, int* indicesBufData, int* detectionsData);
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+        if (isPriorsPerImg) {
+            ppriors += varianceEncodedInTarget ? (n * priorsNum * priorSize) : (2 * n * priorsNum * priorSize);
+            priorVariances += varianceEncodedInTarget ? 0 : (2 * n * priorsNum * priorSize);
+        }
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+        int detectionsTotal = 0;
+        for (int c = 0; c < classesNum; ++c) {
+            detectionsTotal += detectionsData[n * classesNum + c];
         }
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+                int pr = confIndicesClassMap[j].second.second;
+                int *pindices = indicesData + n * classesNum * priorsNum + cls * priorsNum;
+                pindices[detectionsData[n * classesNum + cls]] = pr;
+                detectionsData[n*classesNum + cls]++;
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+    if (!normalized) {
+        int num = 0;
+        for (; num < priorsNum; ++num) {
+            float imgId = priorData[num * priorSize + 0];
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+    if (withAddBoxPred) {
+        for (int n = 0; n < imgNum; ++n) {
+            for (int p = 0; p < priorsNum; ++p) {
+                if (ARMConfData[n*priorsNum*2 + p * 2 + 1] < objScore) {
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-                    ymin = (std::max)(0.0f, (std::min)(1.0f, ymin));
-                    xmax = (std::max)(0.0f, (std::min)(1.0f, xmax));
-                    ymax = (std::max)(0.0f, (std::min)(1.0f, ymax));
+void MKLDNNDetectionOutputNode::confReorderDense(const float *confData, const float *ARMConfData, float *reorderedConfData) {
```
done, thank you.

chenhu-wang commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+        }
+
+        if (isShareLoc) {
+            const float *ploc = locData + n * 4 * priorsNum;
```
done, thank you.

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -126,480 +153,755 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-        if (_conf_data[idx1] > _conf_data[idx2]) return true;
-        if (_conf_data[idx1] < _conf_data[idx2]) return false;
-        return idx1 < idx2;
+    for (int n = 0; n < imgNum; ++n) {
+        for (int c = 0; c < classesNum; ++c) {
+            for (int p = 0; p < priorsNum; ++p) {
```
I think we could use `parallel_for3d` for these cycles

chenhu-wang replied:
done, thank you.

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -125,480 +152,753 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-        // marker at end of boxes list
-        dst_data[count * DETECTION_SIZE + 0] = -1;
+    // withAddBoxPred is false
+    if (permuteKernel_) {
```
Just don't see have the condition can be false. `permuteKernel_` is initialized inside createPrimitive w/o  any condition.

chenhu-wang replied:
Added conditions in createPrimitive(). Thanks!

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -125,480 +152,753 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+                    if (conf > confidenceThreshold) {
+                        int idx = offH + c * confInfoLen;
+                        reorderedConfData[idx + p] = conf;
+                        mtx.lock();
```
Have you tried to avoid mutex here? Like changing order of loops... Is this part of code bottleneck at all?

chenhu-wang replied:
Each prior in parallel process have the possibility to change the statistic info for all classes. It's hard to segment to safe access. Meanwhile, here mutex(sparsity lock check) does not introduce noticeable perf penalty given here is still memory intensive case, although already mitigated dramatically than dense case. Thank you!

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -125,480 +152,753 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+            if (isShareLoc && confInfoV[p] == -1) {
+                return;
+            }
+            if (!isShareLoc && !decreaseClassId && confInfoH[priorsNum] == 0) {
```
This conditions seems independent from the loop iteration. Does it makes sense to extract it from the loop?

chenhu-wang replied:
Extracted out of parallel and act as a check if continue! thank you!

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp:
```diff
@@ -125,480 +152,753 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-            indices[count] = i;
-            count++;
-        }
+inline void MKLDNNDetectionOutputNode::topk(const int *indicesIn, int *indicesOut, const float *conf, int n, int k) {
```
Don't really think we can overtake partial_sort_copy but still does it make sense to try optimized TopK code @xuchen-intel has just implemented?

chenhu-wang replied:
I will discuss if current topk can be apply to this case with xuchen.  If supported, I will have a compare the perf. Anyway, this will be done in a separate PR after topk is finalized. Thank you!

dmitry-gorokhov reviewed (approved):
In general LGTM. Rest comments are minor and can be applied in separate PR if required

dmitry-gorokhov commented:
@chenhu-wang Could you please rebase the PR on latest master to get fresh CI?

dmitry-gorokhov commented:
@chenhu-wang Please rebase on latest master and I will merge the PR

chenhu-wang commented:
> @chenhu-wang Please rebase on latest master and I will merge the PR

@dmitry-gorokhov  rebased! thanks

chenhu-wang commented:
> > @chenhu-wang Please rebase on latest master and I will merge the PR
> 
> @dmitry-gorokhov rebased! thanks

@dmitry-gorokhov , rebased and CI is green now, thank you!
